File: backend/app/models/factor.py
"""Unified Pydantic models for factor construction, evaluation, and backtest.

FactorConfig is the single entry-point config used across compute, eval,
and backtest layers. Each layer adds its own config (EvalConfig, BtConfig)
and produces its own result type (FactorMatrix, FactorReport, BacktestReport).
"""

from __future__ import annotations

import datetime
from enum import Enum
from typing import Literal

from pydantic import BaseModel, Field


# ── Enums ────────────────────────────────────────────────────────────────────

class StockPool(str, Enum):
    HS300 = "hs300"
    ZZ500 = "zz500"
    ZZ800 = "zz800"
    ZZ1000 = "zz1000"
    ZZ_QUANZHI = "zz_quanzhi"


class FactorDirection(str, Enum):
    ASC = "asc"    # smaller factor value is better
    DESC = "desc"  # larger factor value is better


class ICMethod(str, Enum):
    PEARSON = "pearson"
    SPEARMAN = "spearman"


class OutlierHandling(str, Enum):
    NONE = "none"
    WINSORIZE = "winsorize"
    STANDARDIZE = "standardize"


class RebalancePeriod(str, Enum):
    DAILY = "daily"
    WEEKLY = "weekly"
    MONTHLY = "monthly"


class PortfolioType(str, Enum):
    LONG_ONLY = "long_only"
    LONG_SHORT_I = "long_short_i"
    LONG_SHORT_II = "long_short_ii"


# ── Unified Factor Config ─────────────────────────────────────────────────────

class FactorConfig(BaseModel):
    """Single config reused across compute, evaluation, and backtest layers."""
    expression: str = Field(..., description="Factor expression, e.g. '(1/PE_TTM) + ROE'")
    stock_pool: StockPool = Field(..., description="Stock universe")
    start_date: datetime.date
    end_date: datetime.date
    benchmark: str = Field("000300.SH", description="Benchmark symbol")
    direction: FactorDirection = Field(FactorDirection.DESC, description="Factor direction")


# ── Evaluation Config ────────────────────────────────────────────────────────

class EvalConfig(BaseModel):
    """Parameters specific to factor evaluation / IC analysis."""
    ic_method: ICMethod = Field(ICMethod.SPEARMAN)
    group_count: int = Field(5, ge=2, le=20)
    outlier_handling: OutlierHandling = Field(OutlierHandling.WINSORIZE)
    industry_neutralization: bool = False
    include_st: bool = False
    include_new: bool = True


# ── Backtest Config ──────────────────────────────────────────────────────────

class BtConfig(BaseModel):
    """Parameters specific to factor quantile backtest."""
    rebalance_period: RebalancePeriod = Field(RebalancePeriod.MONTHLY)
    fee_rate: float = Field(0.001, ge=0, le=0.05)
    slippage: float = Field(0.001, ge=0, le=0.05)
    filter_limit_up: bool = True
    portfolio_type: PortfolioType = Field(PortfolioType.LONG_ONLY)


# ── Result Types ─────────────────────────────────────────────────────────────

class ICPoint(BaseModel):
    date: datetime.date
    value: float


class IndustryIC(BaseModel):
    industry: str
    value: float


class TurnoverPoint(BaseModel):
    date: datetime.date
    min_quantile: float
    max_quantile: float


class DecayPoint(BaseModel):
    lag: int
    min_quantile: float
    max_quantile: float


class StockFactorValue(BaseModel):
    symbol: str
    name: str
    value: float


class FactorReport(BaseModel):
    """6-module factor analysis report."""
    ic_series: list[ICPoint] = Field(default_factory=list)
    industry_ic: list[IndustryIC] = Field(default_factory=list)
    turnover: list[TurnoverPoint] = Field(default_factory=list)
    signal_decay: list[DecayPoint] = Field(default_factory=list)
    top20: list[StockFactorValue] = Field(default_factory=list)
    bottom20: list[StockFactorValue] = Field(default_factory=list)
    update_date: datetime.date


class BacktestMetrics(BaseModel):
    total_return: float
    annual_return: float
    sharpe: float
    max_drawdown: float
    alpha: float
    beta: float
    ir: float


class NAVPoint(BaseModel):
    date: datetime.date
    value: float


class BacktestReport(BaseModel):
    """Unified backtest output."""
    nav_series: list[NAVPoint] = Field(default_factory=list)
    benchmark_series: list[NAVPoint] = Field(default_factory=list)
    metrics: BacktestMetrics | None = None
    logs: list[str] = Field(default_factory=list)


# ── Board Types ──────────────────────────────────────────────────────────────

class BoardQuery(BaseModel):
    """Query parameters for factor board."""
    categories: list[str] | None = None  # None = all
    stock_pool: StockPool = StockPool.ZZ500
    period: Literal["3m", "1y", "3y", "10y"] = "3y"
    portfolio_type: PortfolioType = PortfolioType.LONG_ONLY
    fee_config: Literal["none", "commission_stamp", "commission_stamp_slippage"] = "none"
    filter_limit_up: bool = True
    sort_by: str = "ic_mean"
    sort_order: Literal["asc", "desc"] = "desc"
    page: int = Field(1, ge=1)
    page_size: int = Field(20, ge=1, le=100)


class BoardRow(BaseModel):
    """Single row in factor board table."""
    factor_name: str
    category: str
    min_quantile_excess_return: float
    max_quantile_excess_return: float
    min_quantile_turnover: float
    max_quantile_turnover: float
    ic_mean: float
    ir: float


class BoardResponse(BaseModel):
    rows: list[BoardRow]
    total: int
    page: int
    page_size: int


# ── Factor Template ──────────────────────────────────────────────────────────

class TemplateType(str, Enum):
    FINANCIAL = "financial"
    TECHNICAL = "technical"
    CUSTOM_OPERATOR = "custom_operator"
    CUSTOM_BASE = "custom_base"


class FactorTemplate(BaseModel):
    """Template for creating a new factor."""
    id: str
    type: TemplateType
    name: str
    description: str
    preset_expression: str
    preset_params: dict = Field(default_factory=dict)
    category: str


# ── Factor CRUD ──────────────────────────────────────────────────────────────

class FactorCreate(BaseModel):
    name: str
    expression: str
    stock_pool: StockPool = StockPool.HS300
    category: str | None = None
    description: str | None = None
    params: dict = Field(default_factory=dict)


class FactorUpdate(BaseModel):
    name: str | None = None
    expression: str | None = None
    stock_pool: StockPool | None = None
    category: str | None = None
    description: str | None = None
    params: dict | None = None


class FactorResponse(BaseModel):
    id: int
    name: str
    expression: str
    stock_pool: StockPool
    category: str | None
    description: str | None
    params: dict
    created_at: str
    updated_at: str


class ValidateRequest(BaseModel):
    expression: str
    stock_pool: StockPool = StockPool.HS300
    date: datetime.date | None = None


class ValidateResponse(BaseModel):
    valid: bool
    error: str | None = None
    preview_rows: list[dict] | None = None

File: frontend/src/types/factor.ts
/** Unified factor config — matches backend FactorConfig Pydantic model */
export interface FactorConfig {
  expression: string
  stock_pool: string
  start_date: string
  end_date: string
  benchmark?: string
  direction?: FactorDirection
}

export type StockPool =
  | 'hs300'
  | 'zz500'
  | 'zz800'
  | 'zz1000'
  | 'zz_quanzhi'
  | 'chinext'
  | 'chinext50'
  | 'chinext_composite'
  | 'star50'
  | 'star100'
export type FactorDirection = 'asc' | 'desc'

/** Evaluation config */
export interface EvalConfig {
  ic_method?: 'pearson' | 'spearman'
  group_count?: number
  outlier_handling?: 'none' | 'winsorize' | 'standardize'
  industry_neutralization?: boolean
  include_st?: boolean
  include_new?: boolean
}

/** Backtest config */
export interface BtConfig {
  rebalance_period?: 'daily' | 'weekly' | 'monthly'
  fee_rate?: number
  slippage?: number
  filter_limit_up?: boolean
  portfolio_type?: 'long_only' | 'long_short_i' | 'long_short_ii'
}

/** Factor template */
export interface FactorTemplate {
  id: string
  type: 'financial' | 'technical' | 'custom_operator' | 'custom_base'
  name: string
  description: string
  preset_expression: string
  preset_params: Record<string, string>
  category: string
}

/** Factor analysis report — 6 modules */
export interface FactorReport {
  ic_series: ICPoint[]
  industry_ic: IndustryIC[]
  turnover: TurnoverPoint[]
  signal_decay: DecayPoint[]
  top20: StockFactorValue[]
  bottom20: StockFactorValue[]
  update_date: string
}

export interface ICPoint {
  date: string
  value: number
}

export interface IndustryIC {
  industry: string
  value: number
}

export interface TurnoverPoint {
  date: string
  min_quantile: number
  max_quantile: number
}

export interface DecayPoint {
  lag: number
  min_quantile: number
  max_quantile: number
}

export interface StockFactorValue {
  symbol: string
  name: string
  value: number
}

/** Backtest report */
export interface BacktestReport {
  nav_series: NAVPoint[]
  benchmark_series: NAVPoint[]
  metrics: BacktestMetrics | null
  logs: string[]
}

export interface NAVPoint {
  date: string
  value: number
}

export interface BacktestMetrics {
  total_return: number
  annual_return: number
  sharpe: number
  max_drawdown: number
  alpha: number
  beta: number
  ir: number
}

/** Board query & response */
export interface BoardQuery {
  categories?: string[]
  factor_groups?: string[]
  factor_keyword?: string
  stock_pool?: string
  period?: '3m' | '1y' | '3y' | '10y'
  start_date?: string | null
  end_date?: string | null
  portfolio_type?: 'long_only' | 'long_short_i' | 'long_short_ii'
  fee_config?: 'none' | 'commission_stamp' | 'commission_stamp_slippage'
  fee_rate?: number
  stamp_tax_rate?: number
  transfer_fee_rate?: number
  slippage?: number
  filter_limit_up?: boolean
  pool_membership_mode?: 'static_latest' | 'point_in_time' | 'union'
  factor_value_params_hashes?: Record<string, string>
  sort_by?: string
  sort_order?: 'asc' | 'desc'
  page?: number
  page_size?: number
}

export interface BoardRow {
  factor_name: string
  display_name?: string | null
  description?: string | null
  source?: string | null
  factor_group?: string | null
  factor_group_display_name?: string | null
  category: string
  coverage_min_date?: string | null
  coverage_max_date?: string | null
  coverage_total_rows?: number
  coverage_symbol_count?: number
  coverage_date_count?: number
  coverage_status?: 'covered' | 'partial' | 'empty' | 'unknown'
  latest_run_id?: string | null
  latest_run_at?: string | null
  latest_ic_mean?: number | null
  latest_icir?: number | null
  latest_long_short_return?: number | null
  latest_max_drawdown?: number | null
  latest_turnover?: number | null
  latest_active_symbol_count?: number | null
  min_quantile_excess_return: number
  max_quantile_excess_return: number
  min_quantile_turnover: number
  max_quantile_turnover: number
  ic_mean: number
  ir: number
}

export interface BoardResponse {
  rows: BoardRow[]
  total: number
  page: number
  page_size: number
}

/** Validation */
export interface ValidateRequest {
  expression: string
  stock_pool?: StockPool
  date?: string
}

export interface ValidateResponse {
  valid: boolean
  error?: string | null
  preview_rows?: Record<string, unknown>[] | null
}
